[SE-0365] Allow implicit self in inner functions in [weak self] closures, like with [self] closures

calda opened the pull request:
This PR fixes an issue in the implementation of [SE-0365](https://github.com/apple/swift-evolution/blob/main/proposals/0365-implicit-self-weak-capture.md), where implicit self was disallowed inside inner function declarations:

```swift
doWorkEscaping { [weak self] in
  guard let self else { return }
  func innerFunction() {
    // Before: Error: call to method 'test' in closure requires explicit use of 'self' to make capture semantics explicit
    // Now allowed without an error
    test()
  }
}
``` 

The intent of SE-0365 is to mirror the behavior of SE-0269 with `[self]` captures. This usage is allowed by SE-0269, so it should be allowed by SE-0365 as well:

```swift
doWorkEscaping { [self] in
  func innerFunction() {
    test() // ✅
  }
}
``` 

I [confirmed on the forums](https://forums.swift.org/t/why-does-se-0269-have-different-rules-for-inner-closures-vs-inner-functions/64334) that this was an intentional design choice of SE-0269.

Please review: @xedin

calda committed:
Allow implicit self in inner functions in [weak self] closures, like with [self] closures

xedin commented on lib/AST/UnqualifiedLookup.cpp:
```diff
@@ -401,19 +401,68 @@ bool implicitSelfReferenceIsUnwrapped(const ValueDecl *selfDecl,
+
+      else if (auto OSP = dyn_cast<OptionalSomePattern>(pattern)) {
+        if (auto subPattern = OSP->getSubPattern()) {
+          if (subPattern->getBoundName() == Ctx.Id_self) {
```
Nit: this could be simplified down to assignment.

xedin commented on lib/AST/UnqualifiedLookup.cpp:
```diff
@@ -401,19 +401,68 @@ bool implicitSelfReferenceIsUnwrapped(const ValueDecl *selfDecl,
+      condDRE = DRE;
+    }
+
+    if (auto LE = dyn_cast<LoadExpr>(cond.getInitializer())) {
```
Hm, maybe we should use getSemanticsProvidingExpr here instead of dealing with LoadExpr directly? Should be able to simply all these checks into one

xedin replied:
We cannot lookup through LoadExpr but I think we still need to use getSemanticsProvidingExpr afterwards maybe in case decl ref was wrapped in parens or something? I’d structure the code in a way where it checks for load, gets semantic expr, checks for DeclRefExpr, WDYT?

xedin reviewed (commented):
Thank you! I left a few comments inline, I would be great to get a few cases with more nesting and i.e. inits and subscripts involved (both valid and invalid)

calda committed:
Clean up implementation, add more extensive tests

calda commented:
Thanks for the review @xedin -- I incorporated your suggestions and added more extensive tests in https://github.com/apple/swift/pull/65211/commits/6670639e9bb233400fd3f18dc54d384c14e7b661

xedin commented on lib/AST/UnqualifiedLookup.cpp:
```diff
@@ -401,19 +401,66 @@ bool implicitSelfReferenceIsUnwrapped(const ValueDecl *selfDecl,
+      return false;
+    }
+
+    return condDRE->getDecl()->getName().isSimpleName(Ctx.Id_self);
```
And one more small thing - let’s check whether decl has a name too.

calda committed:
Use getSemanticsProvidingExpr() and hasName()

xedin reviewed (approved):
Thank you!

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

calda commented:
The one issue in the Release source compat suite is unrelated:

> build. MovieSwift, 5.0, 237010, MovieSwift, generic/platform=iOS
> This project built successfully, but it was expected to fail.  https://github.com/apple/swift/issues/65203
